File: src/mocks/seed.ts
import type {
  ApiKey,
  Approval,
  AuditLog,
  ChatSession,
  KnowledgeBase,
  KnowledgeDocument,
  ModelConfig,
  User,
} from '../types/domain';

export const defaultUser: User = {
  id: 'u-1001',
  name: '张工',
  department: '水声技术部',
  role: '科研人员',
  ip: '10.12.8.105',
};

export const knowledgeBases: KnowledgeBase[] = [
  {
    id: 'kb-acoustic',
    name: '水声基础理论库',
    department: '水声技术部',
    level: '公开级',
    fileCount: 1250,
    status: '已索引',
    updatedAt: '2026-06-20',
    role: '管理员',
    type: '部门库',
  },
  {
    id: 'kb-project',
    name: '项目报告库',
    department: '项目管理办公室',
    level: '内部级',
    fileCount: 342,
    status: '索引中',
    updatedAt: '2026-06-22',
    role: '查看者',
    type: '授权库',
  },
  {
    id: 'kb-code',
    name: '软件代码规范库',
    department: '软件研发部',
    level: '公开级',
    fileCount: 56,
    status: '已索引',
    updatedAt: '2026-06-15',
    role: '查看者',
    type: '部门库',
  },
  {
    id: 'kb-personal',
    name: '个人文档速查',
    department: '个人',
    level: '私有',
    fileCount: 12,
    status: '已索引',
    updatedAt: '刚才',
    role: '所有者',
    type: '个人库',
  },
];

export const documents: KnowledgeDocument[] = [
  {
    id: 'doc-001',
    knowledgeBaseId: 'kb-acoustic',
    title: '水声学原理第四章：海洋中的声传播',
    fileName: '水声学原理-第四章.pdf',
    status: '已入库',
    securityResult: '通过',
    applicant: '系统初始化',
    submittedAt: '2026-06-20 09:30',
    summary: '解释声速剖面、SOFAR 声道、汇聚区和影区对探测距离的影响。',
  },
  {
    id: 'doc-002',
    knowledgeBaseId: 'kb-project',
    title: '海试项目批次 2025-A 数据总报告',
    fileName: '海试项目批次2025-A数据总报告.pdf',
    status: '已入库',
    securityResult: '通过',
    applicant: '王主任',
    submittedAt: '2026-06-18 15:20',
    summary: '记录复杂海况下声呐平台探测距离变化和跃层影响。',
  },
  {
    id: 'doc-003',
    knowledgeBaseId: 'kb-acoustic',
    title: '声呐阵列信号处理算法预研报告',
    fileName: '声呐阵列信号处理算法预研报告.docx',
    status: '待审核',
    securityResult: '通过',
    applicant: '张工',
    submittedAt: '2026-06-29 14:10',
    summary: '申请入库的算法预研报告，等待知识库管理员复核。',
  },
];

export const approvals: Approval[] = [
  {
    id: 'ap-001',
    type: '文件入库',
    applicant: '张工',
    target: '声呐阵列信号处理算法预研报告.docx',
    status: '待审批',
    risk: '中风险',
    createdAt: '2026-06-29 14:10',
    relatedDocumentId: 'doc-003',
  },
  {
    id: 'ap-002',
    type: '知识库授权',
    applicant: '李研究员',
    target: '申请访问：项目报告库',
    status: '待审批',
    risk: '无风险',
    createdAt: '2026-06-29 09:30',
  },
];

export const chatSessions: ChatSession[] = [
  {
    id: 'chat-001',
    title: '声速剖面对探测距离的影响分析',
    model: 'Qwen3-30B-A3B-w8a8',
    updatedAt: '2026-06-29 14:30',
    messages: [
      {
        id: 'msg-001',
        role: 'user',
        content: '请基于水声知识库说明声速剖面对声呐探测距离的影响。',
        createdAt: '2026-06-29 14:28',
      },
      {
        id: 'msg-002',
        role: 'assistant',
        model: 'Qwen3-30B-A3B-w8a8',
        content:
          '声速剖面对声呐探测距离具有决定性影响。海水温度、盐度和压力会随深度变化，导致声线折射。表面声道会延长近海面传播距离，深海声道可形成远距离低损耗传播，汇聚区会在远距离形成能量增强，而强跃层可能制造影区，使实际探测距离明显低于标称值。',
        createdAt: '2026-06-29 14:29',
        citations: [
          {
            id: 'cite-001',
            documentId: 'doc-001',
            knowledgeBaseId: 'kb-acoustic',
            title: '水声学原理第四章：海洋中的声传播',
            knowledgeBaseName: '水声基础理论库',
            similarity: 98,
            excerpt: '当声源位于声速极小值轴附近时，声能会被捕获在声道内传播，边界散射损耗较小。',
          },
          {
            id: 'cite-002',
            documentId: 'doc-002',
            knowledgeBaseId: 'kb-project',
            title: '海试项目批次 2025-A 数据总报告',
            knowledgeBaseName: '项目报告库',
            similarity: 85,
            excerpt: '夏季强跃层导致影区现象明显，被动平台探测距离较冬季同等海况下降超过 60%。',
          },
        ],
      },
    ],
  },
];

export const auditLogs: AuditLog[] = [
  {
    id: 'aud-001',
    time: '2026-06-29 14:30:12',
    user: '张工',
    role: '科研人员',
    action: '智能问答',
    resource: '水声基础理论库',
    ip: '10.12.8.105',
    risk: 'normal',
    detail: '生成回答并引用 2 条知识库依据。',
  },
  {
    id: 'aud-002',
    time: '2026-06-29 14:10:30',
    user: '张工',
    role: '科研人员',
    action: '文件上传',
    resource: '声呐阵列信号处理算法预研报告.docx',
    ip: '10.12.8.105',
    risk: 'warning',
    detail: '文件通过初筛，已进入入库审批。',
  },
];

export const modelConfigs: ModelConfig[] = [
  { id: 'm-001', name: 'Qwen3-30B-A3B-w8a8', type: '通用大语言模型', status: '正常', isDefault: true, endpoint: '192.168.10.101:8000/v1' },
  { id: 'm-002', name: 'Qwen 3.5', type: '通用大语言模型', status: '正常', isDefault: false, endpoint: '10.12.1.201:8080/v1' },
  { id: 'm-003', name: 'DeepSeek V4', type: '深度推理模型', status: '正常', isDefault: false, endpoint: '10.12.2.100:8000/v1' },
  { id: 'm-004', name: 'DeepSeek R1', type: '备用推理模型', status: '已下线', isDefault: false, endpoint: '-' },
];

export const apiKeys: ApiKey[] = [
  { id: 'key_live_9x8a2b', name: '科研平台后端调用', caller: '平台A-资源服务', expiry: '2026-12-31', limit: '1000/天', status: '正常', createdAt: '2026-01-10' },
  { id: 'key_test_1f3c5d', name: '内部工具测试', caller: '张工（个人分配）', expiry: '2026-06-30', limit: '100/天', status: '正常', createdAt: '2026-06-01' },
];

File: backend/app/seed.py
"""本地演示数据初始化。

启动应用时调用，确保角色、用户、模型、知识库等基础数据存在。
"""

import hashlib
from datetime import datetime
from pathlib import Path
from uuid import uuid4

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.config import get_settings
from app.models import (
    Approval,
    ApprovalStatus,
    ApprovalType,
    ApiKey,
    AuditLog,
    AuditRisk,
    ChatMessage,
    ChatRole,
    ChatSession,
    DocumentStatus,
    KnowledgeBase,
    KnowledgeDocument,
    KnowledgeDocumentChunk,
    LLMTask,
    KnowledgeLevel,
    KnowledgeRole,
    KnowledgeStatus,
    KnowledgeType,
    ModelConfig,
    ModelStatus,
    Notification,
    RiskLevel,
    Role,
    RoleName,
    SecurityResult,
    User,
)


def now_text() -> str:
    """返回数据库中统一使用的时间字符串。"""
    return datetime.now().strftime("%Y-%m-%d %H:%M:%S")


def new_id(prefix: str) -> str:
    return f"{prefix}-{uuid4().hex[:10]}"


def hash_secret(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()


def ensure_default_passwords(db: Session) -> None:
    for user in db.scalars(select(User)).all():
        if not user.password_hash:
            user.password_hash = hash_secret("123456")


def ensure_platform_records(db: Session) -> None:
    if db.get(ApiKey, "key-demo"):
        return
    db.add(
        ApiKey(
            id="key-demo",
            name="内部系统联调密钥",
            caller="科研协同门户",
            key_hash=hash_secret("sk-lp-demo"),
            scopes="chat:invoke,kb:search",
            expiry="长期有效",
            limit="1000/day",
            status="正常",
            created_at=now_text(),
        )
    )


def ensure_merged_default_roles(db: Session) -> None:
    role_by_name = {role.name: role for role in db.scalars(select(Role)).all()}
    researcher = role_by_name.get(RoleName.RESEARCHER)
    kb_admin = role_by_name.get(RoleName.KB_ADMIN)
    if researcher:
        researcher.description = "科研人员（合并普通用户），可使用问答、写作、办公、会议、知识库、上传、代码助手和我的申请"
        for user in db.scalars(select(User).where(User.role_id.in_(["role-normal", "role-researcher"]))).all():
            user.role_id = researcher.id
    if kb_admin:
        kb_admin.description = "知识库管理员（合并安全审计员），可使用知识库管理、入库审核、待审批、审计和报表"
        for user in db.scalars(select(User).where(User.role_id.in_(["role-kb-admin", "role-auditor"]))).all():
            user.role_id = kb_admin.id
    notifications = [
        Notification(id="msg-seed-001", user_id="u-1001", title="知识库索引完成", content="真实种子文档已完成本地索引，可在智能问答中引用。", category="kb", created_at=now_text()),
        Notification(id="msg-seed-002", user_id=None, title="模型网关已配置", content="默认模型端点使用局域网 GLM-5.1。", category="ops", created_at=now_text()),
    ]
    for notification in notifications:
        if db.get(Notification, notification.id) is None:
            db.add(notification)
    if db.get(LLMTask, "task-seed-001") is None:
        db.add(
            LLMTask(
                id="task-seed-001",
                user_id="u-1001",
                task_type="writing",
                title="平台功能说明草稿",
                input_text="说明当前平台能力",
                output_text="平台已接入聊天、知识库、审批、审计和模型网关等核心能力。",
                model=get_settings().llm_model_id,
                status="completed",
                input_tokens=8,
                output_tokens=24,
                created_at=now_text(),
            )
        )


def sync_runtime_model_config(db: Session) -> None:
    settings = get_settings()
    model = db.get(ModelConfig, "m-001")
    if model is None:
        model = ModelConfig(
            id="m-001",
            name=settings.llm_model_id,
            type="通用大语言模型",
            status=ModelStatus.NORMAL,
            is_default=True,
            endpoint=settings.llm_api_base_url,
        )
        db.add(model)
    else:
        model.name = settings.llm_model_id
        model.status = ModelStatus.NORMAL
        model.is_default = True
        model.endpoint = settings.llm_api_base_url
    for other in db.scalars(select(ModelConfig).where(ModelConfig.id != "m-001")):
        other.is_default = False


def sync_runtime_model_config(db: Session) -> None:
    settings = get_settings()
    runtime_models = [
        ("m-001", settings.llm_model_id, "因果推理模型", settings.llm_api_base_url, True),
        ("m-vl-001", settings.vl_llm_model_id, "视觉语言因果推理模型", settings.vl_llm_api_base_url, False),
    ]
    active_ids = {item[0] for item in runtime_models}
    for model_id, name, model_type, endpoint, is_default in runtime_models:
        model = db.get(ModelConfig, model_id)
        if model is None:
            model = ModelConfig(id=model_id, name=name, type=model_type, status=ModelStatus.NORMAL, is_default=is_default, endpoint=endpoint)
            db.add(model)
        else:
            model.name = name
            model.type = model_type
            model.status = ModelStatus.NORMAL
            model.is_default = is_default
            model.endpoint = endpoint
    for other in db.scalars(select(ModelConfig).where(ModelConfig.id.not_in(active_ids))):
        other.status = ModelStatus.OFFLINE
        other.is_default = False


def sync_real_seed_documents(db: Session) -> None:
    root = Path(__file__).resolve().parents[1]
    seed_documents = [
        {
            "id": "doc-001",
            "knowledge_base_id": "kb-acoustic",
            "title": "NOAA Underwater Sound Propagation Notes",
            "file_name": "noaa_underwater_sound_summary.md",
            "path": root / "storage" / "seed_documents" / "noaa_underwater_sound_summary.md",
            "summary": "Based on public NOAA educational material about underwater sound, sound-speed profiles, ducts, thermoclines, and sonar range variation.",
            "chunk_id": "chunk-seed-acoustic-001",
            "page_label": "NOAA summary",
            "submitted_at": "2026-06-20 09:30",
        },
        {
            "id": "doc-002",
            "knowledge_base_id": "kb-code",
            "title": "Python PEP 8 Operational Summary",
            "file_name": "python_pep8_operational_summary.md",
            "path": root / "storage" / "seed_documents" / "python_pep8_operational_summary.md",
            "summary": "Based on the official Python PEP 8 style guide for code layout, naming, imports, comments, and review consistency.",
            "chunk_id": "chunk-seed-code-001",
            "page_label": "PEP 8 summary",
            "submitted_at": "2026-06-18 15:20",
        },
        {
            "id": "doc-003",
            "knowledge_base_id": "kb-project",
            "title": "NIST AI Risk Management Framework Summary",
            "file_name": "nist_ai_rmf_summary.md",
            "path": root / "storage" / "seed_documents" / "nist_ai_rmf_summary.md",
            "summary": "Based on the public NIST AI RMF 1.0 page, covering govern, map, measure, and manage risk functions.",
            "chunk_id": "chunk-seed-project-001",
            "page_label": "NIST AI RMF summary",
            "submitted_at": "2026-06-19 10:00",
        },
    ]

    for item in seed_documents:
        path = item["path"]
        content = path.read_text(encoding="utf-8")
        content_hash = hashlib.sha256(path.read_bytes()).hexdigest()

        document = db.get(KnowledgeDocument, item["id"])
        if document is None:
            document = KnowledgeDocument(
                id=item["id"],
                knowledge_base_id=item["knowledge_base_id"],
                title=item["title"],
                file_name=item["file_name"],
                status=DocumentStatus.INDEXED,
                security_result=SecurityResult.PASSED,
                applicant="System Seed",
                submitted_at=item["submitted_at"],
                summary=item["summary"],
            )
            db.add(document)

        document.knowledge_base_id = item["knowledge_base_id"]
        document.title = item["title"]
        document.file_name = item["file_name"]
        document.status = DocumentStatus.INDEXED
        document.security_result = SecurityResult.PASSED
        document.applicant = "System Seed"
        document.submitted_at = item["submitted_at"]
        document.summary = item["summary"]
        document.storage_path = str(path)
        document.mime_type = "text/markdown"
        document.file_size = path.stat().st_size
        document.content_hash = content_hash
        document.index_status = "indexed"
        document.chunk_count = 1
        document.indexed_at = item["submitted_at"]
        document.index_error = None

        chunk = db.get(KnowledgeDocumentChunk, item["chunk_id"])
        if chunk is None:
            chunk = KnowledgeDocumentChunk(
                id=item["chunk_id"],
                document_id=item["id"],
                knowledge_base_id=item["knowledge_base_id"],
                chunk_index=0,
                text=content,
                page_label=item["page_label"],
                vector_id=f"{item['id']}:0",
                created_at=item["submitted_at"],
            )
            db.add(chunk)
        else:
            chunk.document_id = item["id"]
            chunk.knowledge_base_id = item["knowledge_base_id"]
            chunk.chunk_index = 0
            chunk.text = content
            chunk.page_label = item["page_label"]
            chunk.vector_id = f"{item['id']}:0"
            chunk.created_at = item["submitted_at"]

    for kb_id in {"kb-acoustic", "kb-code", "kb-project"}:
        kb = db.get(KnowledgeBase, kb_id)
        if kb:
            kb.status = KnowledgeStatus.INDEXED
            kb.updated_at = now_text()
            kb.file_count = len(db.scalars(select(KnowledgeDocument).where(KnowledgeDocument.knowledge_base_id == kb_id)).all())

    demo_message = db.get(ChatMessage, "msg-002")
    if demo_message:
        demo_message.citations_json = (
            '[{"id":"chunk-seed-acoustic-001","documentId":"doc-001","knowledgeBaseId":"kb-acoustic",'
            '"title":"NOAA Underwater Sound Propagation Notes","knowledgeBaseName":"水声基础理论库",'
            '"similarity":95,"excerpt":"Sound-speed profiles, ducts, thermoclines, and shadow zones affect sonar detection range."}]'
        )


def seed_demo_data(db: Session) -> None:
    """幂等写入演示数据；已存在的数据不会重复创建。"""
    settings = get_settings()
    if db.scalar(select(Role).limit(1)):
        ensure_merged_default_roles(db)
        sync_runtime_model_config(db)
        sync_real_seed_documents(db)
        ensure_default_passwords(db)
        ensure_platform_records(db)
        db.commit()
        return

    roles = [
        Role(id="role-normal", name=RoleName.NORMAL_USER, description="普通办公与问答用户"),
        Role(id="role-researcher", name=RoleName.RESEARCHER, description="科研人员，可使用问答、知识库和代码助手"),
        Role(id="role-kb-admin", name=RoleName.KB_ADMIN, description="知识库管理员，可审核入库文件"),
        Role(id="role-auth-admin", name=RoleName.AUTH_ADMIN, description="授权管理员，可审批资源访问"),
        Role(id="role-auditor", name=RoleName.AUDITOR, description="安全审计员，可查看审计日志"),
        Role(id="role-ops", name=RoleName.OPS, description="运维账号，可查看模型和运行状态"),
    ]
    db.add_all(roles)
    db.flush()

    role_by_name = {role.name: role for role in roles}
    db.add_all(
        [
            User(id="u-1001", name="张工", department="水声技术部", role_id=role_by_name[RoleName.RESEARCHER].id, ip="10.12.8.105"),
            User(id="u-1002", name="李研究员", department="声呐系统部", role_id=role_by_name[RoleName.NORMAL_USER].id, ip="10.12.8.106"),
            User(id="u-1882", name="赵审计员", department="安全合规部", role_id=role_by_name[RoleName.AUDITOR].id, ip="10.12.8.101"),
            User(id="u-3001", name="授权管理员", department="数据资源管理部", role_id=role_by_name[RoleName.AUTH_ADMIN].id, ip="10.12.8.120"),
            User(id="u-9001", name="运维账号", department="信息中心", role_id=role_by_name[RoleName.OPS].id, ip="10.12.8.200"),
        ]
    )
    db.flush()
    ensure_merged_default_roles(db)

    db.add_all(
        [
            ModelConfig(id="m-001", name=settings.llm_model_id, type="通用大语言模型", status=ModelStatus.NORMAL, is_default=True, endpoint=settings.llm_api_base_url),
            ModelConfig(id="m-002", name="Qwen 3.5", type="通用大语言模型", status=ModelStatus.NORMAL, is_default=False, endpoint="10.12.1.201:8080/v1"),
            ModelConfig(id="m-003", name="DeepSeek V4", type="深度推理模型", status=ModelStatus.NORMAL, is_default=False, endpoint="10.12.2.100:8000/v1"),
            ModelConfig(id="m-004", name="DeepSeek R1", type="备用推理模型", status=ModelStatus.OFFLINE, is_default=False, endpoint="-"),
        ]
    )

    db.add_all(
        [
            KnowledgeBase(id="kb-acoustic", name="水声基础理论库", department="水声技术部", level=KnowledgeLevel.PUBLIC, file_count=0, status=KnowledgeStatus.INDEXED, updated_at=now_text(), role=KnowledgeRole.ADMIN, type=KnowledgeType.DEPARTMENT),
            KnowledgeBase(id="kb-project", name="项目与治理资料库", department="项目管理办公室", level=KnowledgeLevel.INTERNAL, file_count=0, status=KnowledgeStatus.INDEXED, updated_at=now_text(), role=KnowledgeRole.VIEWER, type=KnowledgeType.AUTHORIZED),
            KnowledgeBase(id="kb-code", name="软件代码规范库", department="软件研发部", level=KnowledgeLevel.PUBLIC, file_count=0, status=KnowledgeStatus.INDEXED, updated_at=now_text(), role=KnowledgeRole.VIEWER, type=KnowledgeType.DEPARTMENT),
            KnowledgeBase(id="kb-personal", name="个人文档速查", department="个人", level=KnowledgeLevel.PRIVATE, file_count=0, status=KnowledgeStatus.NOT_INDEXED, updated_at=now_text(), role=KnowledgeRole.OWNER, type=KnowledgeType.PERSONAL),
        ]
    )

    sync_real_seed_documents(db)

    db.add_all(
        [
            Approval(id="ap-001", type=ApprovalType.DOCUMENT_INDEX, applicant="张工", target="待入库研究笔记.md", status=ApprovalStatus.PENDING, risk=RiskLevel.MEDIUM, created_at=now_text()),
            Approval(id="ap-002", type=ApprovalType.KB_AUTH, applicant="李研究员", target="申请访问：项目与治理资料库", status=ApprovalStatus.PENDING, risk=RiskLevel.NONE, created_at=now_text()),
        ]
    )

    db.add_all(
        [
            AuditLog(id="aud-001", time=now_text(), user="张工", role=RoleName.RESEARCHER, action="智能问答", resource="水声基础理论库", ip="10.12.8.105", risk=AuditRisk.NORMAL, detail="基于真实 Markdown 种子文档生成回答并引用知识库依据。"),
            AuditLog(id="aud-002", time=now_text(), user="系统", role=RoleName.OPS, action="知识库初始化", resource="真实种子文档", ip="127.0.0.1", risk=AuditRisk.NORMAL, detail="已同步 NOAA、Python PEP 8、NIST AI RMF 三份真实来源摘要文档。"),
        ]
    )

    session = ChatSession(id="chat-001", title="声速剖面对探测距离的影响", model=settings.llm_model_id, updated_at=now_text())
    db.add(session)
    db.add_all(
        [
            ChatMessage(id="msg-001", session_id=session.id, role=ChatRole.USER, content="请基于水声知识库说明声速剖面对声呐探测距离的影响。", created_at=now_text()),
            ChatMessage(
                id="msg-002",
                session_id=session.id,
                role=ChatRole.ASSISTANT,
                model=settings.llm_model_id,
                content="声速剖面会改变声线传播路径。温度、盐度和压力共同影响海水声速；表面声道、深海声道和热跃层都可能改变声能分布，从而影响声呐探测距离和盲区。",
                created_at=now_text(),
                citations_json='[{"id":"chunk-seed-acoustic-001","documentId":"doc-001","knowledgeBaseId":"kb-acoustic","title":"NOAA Underwater Sound Propagation Notes","knowledgeBaseName":"水声基础理论库","similarity":95,"excerpt":"Sound-speed profiles, ducts, thermoclines, and shadow zones affect sonar detection range."}]',
            ),
        ]
    )

    ensure_default_passwords(db)
    ensure_platform_records(db)
    db.commit()
